lib/Base/NumpyReader.cpp

```diff
@@ -0,0 +1,302 @@
+  }
+
+  normalizeData(ImageLayout::Unspecified, mean, stddev, imageNormMode, data,
+                dataNpy.shape);
+
+  inputData.reset(ElemKind::FloatTy, dataNpy.shape);
+  inputData.getHandle<>() = data;
+}
+
+void glow::loadNumpyImagesAndPreprocess(
+    const llvm::ArrayRef<std::string> &filenames, Tensor &inputData,
+    ImageNormalizationMode imageNormMode, ImageLayout imageLayout,
+    ImageLayout inputLayout, llvm::ArrayRef<float> mean,
+    llvm::ArrayRef<float> stddev) {
+  DCHECK(!filenames.empty())
+      << "NPY loader: There must be at least one filename in filenames.";
+
+  if (imageLayout == ImageLayout::Unspecified) {
+    return loadUnspecifiedImageAndPreprocess(filenames, inputData,
+                                             imageNormMode, mean, stddev);
+  }
+
+  dim_t numImg = filenames.size();
+
+  // Read each tensor file into a vector of tensors.
+  std::vector<Tensor> tensors(numImg);
+  dim_t batchSize = 0;
+  for (dim_t n = 0; n < numImg; n++) {
+
+    NpyData dataNpy;
+    numpyReader(filenames[n], dataNpy);
+    std::vector<float> data(dataNpy.nvals);
+    convertU8S8ToFloat(dataNpy, data);
+    // Expand 3D to 4D. Supporting 4D only.
+    if (dataNpy.shape.size() == 3) {
+      dataNpy.shape.insert(dataNpy.shape.begin(), 1);
+    }
+    CHECK_EQ(dataNpy.shape.size(), 4)
+        << "NPY loader: Supporting only 3 or 4 dimensions.";
+    normalizeData(imageLayout, mean, stddev, imageNormMode, data,
+                  dataNpy.shape);
+    // Load tensor from the vector obtained from the npy loader.
+    tensors[n].reset(glow::ElemKind::FloatTy, dataNpy.shape);
+    tensors[n].getHandle<>() = data;
+    auto dims0 = tensors[0].dims();
+    auto dims = tensors[n].dims();
+    if (n > 1) {
+      CHECK_EQ(dims0.size(), dims.size())
+          << "NPY loader: Number of dimensions must match.";
+      for (dim_t i = 1; i < dims.size(); i++) {
+        CHECK_EQ(dims0[i], dims[i])
+            << "NPY loader: Non-batch dimensions must match.";
+      }
+    }
+    // Accumulate batch dimension after each dump is loaded.
+    batchSize += tensors[n].dims()[0];
+  }
+  // Input tensor dimensions are now fully known.
+  inputData.reset(ElemKind::FloatTy,
+                  {batchSize, tensors[0].dims()[1], tensors[0].dims()[2],
+                   tensors[0].dims()[3]});
+  auto IIDH = inputData.getHandle<>();
+  // Insert each loaded file (in tensors[] tensors) as the input tensor slices.
+  for (dim_t n = 0, batch = 0, e = tensors.size(); n < e; n++) {
+    Handle<float> H = tensors[n].getHandle<>();
+    IIDH.insertTensors(H, {batch, 0, 0, 0});
+    batch += tensors[n].dims()[0];
+  }
+
+  // Convert to the requested layout.
+  if (inputLayout != imageLayout && inputLayout != ImageLayout::Unspecified) {
+    glow::Tensor transposed;
+    if (imageLayout == ImageLayout::NCHW) {
+      inputData.transpose(&transposed, {0u, 3u, 1u, 2u});
+    } else {
+      inputData.transpose(&transposed, {0u, 2u, 3u, 1u});
```

Comment:
Can we move `0u, 3u, 1u, 2u` and other into a `#define NHWC2NCHW`, and reuse elsewhere in this file? As is done in Graph.h (I'm assuming you didn't want to include that here because of a dependency issue?)

Reply:
I didn't include Graph.h as it seem not to belong here - here I am working with Tensors only and some image processing stuff, thus no need to include anything else.  The transpose formats maybe should be in Tensor.h ? Otherwise, I can replicate here